addon/services/model-proxy.js

```diff
@@ -12,24 +14,145 @@ export default Service.extend({
     return app.lookup('util:model-proxy', { singleton: false });
   },
 
-  _getRealModel(model){
-    if (model && model.get) {
-      return get(model, 'content') || model;
+  _getRealModel(model) {
+    if (model && get(model, 'content')) {
+      return get(model, 'content');
     }
     return model;
   },
 
+  _setupCurrentState(proxy, model) {
+    if (model) {
+      var currentState = EmberObject.create({
+        isEmpty: get(model, 'currentState.isEmpty'),
+        isLoading: get(model, 'currentState.isLoading'),
+        isLoaded: get(model, 'currentState.isLoaded'),
+        isDirty: get(model, 'currentState.isDirty'),
+        isSaving: get(model, 'currentState.isSaving'),
+        isDeleted: get(model, 'currentState.isDeleted'),
+        isNew: get(model, 'currentState.isNew'),
+        isValid: get(model, 'currentState.isValid')
+      });
+    } else {
+      currentState = EmberObject.create({
+        isEmpty: false,
+        isLoading: false,
+        isLoaded: false,
+        isDirty: false,
+        isSaving: false,
+        isDeleted: false,
+        isNew: true,
+        isValid: true
+      });
+    }
+
+    let internalProxy = get(proxy, 'proxy');
+    internalProxy.currentState = currentState;
+    internalProxy._internalModel.currentState = currentState;
+  },
+  _addObserver(proxy, model, propertyName) {
```

Comment:
I think the spacing between functions in this file lacks consistency -- either single space between functions or no spaces (I'm in the space camp).